pycbc/io/hdf.py

```diff
@@ -402,47 +469,84 @@ class SingleDetTriggers(object):
     """
     Provides easy access to the parameters of single-detector CBC triggers.
     """
-    # FIXME: Some of these are optional and should be kwargs.
-    def __init__(self, trig_file, bank_file, veto_file,
-                 segment_name, filter_func, detector, premask=None):
+    def __init__(self, trig_file, detector, bank_file=None, veto_file=None,
+                 segment_name=None, premask=None, filter_rank=None,
+                 filter_threshold=None, chunksize=int(1e6), filter_func=None):
+        """
+        Create a SingleDetTriggers instance
+
+        Parameters
+        ----------
+        trig_file : string or os.pathtype, required
+            HDF file containing trigger information
+
+        detector : string, required
+            The detectior being used, this is used to access the
+            triggers in trig_file
+
+        bank_file: string or os.pathtype, optional
+            hdf file containing template bank information
+
+        veto_file: string or os.pathtype, optional
+            File used to define vetoes
+
+        segment_name : string, optional
+            Segment name being used in the veto_file
+
+        premask : array of indices or boolean, optional
+           Array of used triggers
+
+        filter_rank : string, optional
+            The ranking, as defined by ranking.py to compare to
+            filter_threshold
+
+        filter_threshold: float, required if filter_rank is used
+            Threshold to filter the ranking values
+
+        chunksize : int , default 1e6
+            Size of chunks to read in for the filter_rank / threshold.
+        """
         logging.info('Loading triggers')
         self.trigs_f = HFile(trig_file, 'r')
         self.trigs = self.trigs_f[detector]
+        self.ntriggers = self.trigs['end_time'].size
         self.ifo = detector  # convenience attributes
         self.detector = detector
         if bank_file:
             logging.info('Loading bank')
             self.bank = HFile(bank_file, 'r')
         else:
-            logging.info('No bank file given to SingleDetTriggers')
             # empty dict in place of non-existent hdf file
             self.bank = {}
 
-        if premask is not None:
-            self.mask = premask
-        else:
-            self.mask = np.ones(len(self.trigs['end_time']), dtype=bool)
+        # Apply some masks to start off with - here we should try and apply
+        # them in the order which cuts most things earliest.
+        self.mask = premask
+
+        if filter_rank:
+            assert filter_threshold is not None
+            logging.info("Applying threshold of %.3f on %s",
+                         filter_threshold, filter_rank)
+            idx = self.trigs_f.select(
+                 lambda rank: rank > filter_threshold,
+                 derived={ranking.sngls_ranking_function_dict[filter_rank]:
```

Comment:
That would lose some functionality compared to what is in there at the moment. As it would then restrict us to the ranking functions only. We could use a tuple as the value, which defines the function and the datasets required, so we are not hard-coded only to those

e.g.
```
indices, tuple_out = trig_f.select(
    a_function_which_uses_end_time_and_newsnr_sgveto,
    'end_time',
    'newsnr_sgveto'
    group=opts.detector
    derived={'newsnr_sgveto': (ranking.sngls_ranking_function_dict['newsnr_sgveto'], 
                               ranking.required_datasets['newsnr_sgveto'])
)
```
which would return a tuple with 'end_time' and 'newsnr_sgveto'

Reply:
I have made this change